plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,245 @@
+won't get us much as they aren't created very often.
+-}
+
+processSingleBinding
+    :: Inlining tyname name uni a term m
+    => Binding tyname name uni a
+    -> m (Maybe (Binding tyname name uni a))
+processSingleBinding = \case
+    -- See Note [Inlining various kinds of binding]
+    TermBind a s v@(VarDecl _ n _) rhs -> do
+        maybeRhs' <- maybeAddSubst s n rhs
+        case maybeRhs' of
+            Just rhs' -> pure $ Just $ TermBind a s v rhs'
+            Nothing -> pure Nothing
+    -- Not a strict binding, just process all the subterms
+    b -> Just <$> forMOf bindingSubterms b processTerm
+
+maybeAddSubst
+    :: Inlining tyname name uni a term m
+    => Strictness
+    -> name
+    -> Term tyname name uni a
+    -> m (Maybe (Term tyname name uni a))
+maybeAddSubst s n rhs = do
+    -- Only do PostInlineUnconditional
+    -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+    rhs' <- processTerm rhs
+    doInline <- postInlineUnconditional s rhs'
+    if doInline then do
+        modify (\subst -> extendSubst n (Done rhs') subst)
+        pure Nothing
+    else pure $ Just rhs'
+
+{- Note [Inlining criteria]
+What gets inlined? We don't really care about performance here, so we're really just
+angling to simplify the code without making things worse.
+
+The obvious candidates are tiny things like builtins, variables, or constants.
+We could also consider inlining variables with arbitrary RHSs that are used only
+once, but we don't do that currently.
+-}
+
+{- Note [Inlining and purity]
+When can we inline something that might have effects? We must remember that we often also
+remove a binding that we inline.
+
+For strict bindings, the answer is that we can't: we will delay the effects to the use site,
+so they may happen multiple times (or none). So we can only inline bindings whose RHS is pure.
+
+For non-strict bindings, the effects already happened at the use site, so it's fine to inline it
+unconditionally.
+-}
+
+-- | Should we inline? Should only inline things that won't duplicate work or code.
+-- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+postInlineUnconditional :: Inlining tyname name uni a term m => Strictness -> Term tyname name uni a -> m Bool
+postInlineUnconditional s t = do
+    (strictnessMap, means) <- ask
+    let -- See Note [Inlining criteria]
+        termIsTrivial = trivialTerm t
+        -- See Note [Inlining and purity]
+        strictnessFun = \n' -> Map.findWithDefault NonStrict (n' ^. theUnique) strictnessMap
+        termIsPure = case s of { Strict -> isPure means strictnessFun t; NonStrict -> True; }
+    pure $ termIsTrivial && termIsPure
+
+-- | Is this a an utterly trivial term which might as well be inlined?
+trivialTerm :: Term tyname name uni a -> Bool
+trivialTerm = \case
+    Builtin{} -> True
+    Var{} -> True
+    -- TODO: Should this depend on the size of the constant?
+    Constant{} -> True
+    _ -> False
```

Comment:
Why don't we also add `App t1 t2 -> trivialTerm t1 && trivialTerm t2` ? (pardon my ignorance,  I haven't read the ghc inliner paper yet).

Reply:
If you have `y = f x` and you inline `y` twice, you'll have to compute `f x` twice, which is undesirable.

Reply:
Is it safe though in terms of effects? Also, your example is a cornercase. There can be cases where it can be helpful?

Reply:
Now that I am thinking it more, you are right, it does not make sense to inline the App; unless I guess you plan to inline it only once...

Reply:
> Is it safe though in terms of effects?

Inlining a `let`? In Haskell yes, in Plutus no. So yeah, another reason.

> Also, your example is a cornercase. There can be cases where it makes sense, right?

That function is there for handling the particular corner case of a completely trivial binding. Usage analysis is performed elsewhere and yes, it may result in `y = f x` being inlined (provided `y` is only used once and not under a lambda; although in the case of Plutus that may also change the semantics of a program -- the effects thing that you mentioned).

Reply:
We don't do inlining based on usage analysis. There are cases where we could inline values that are used only once (as Roman says we would need to be careful about effects - but we already only inline strict bindings when they are pure).

We really can't consider multiply used bindings to be a corner case - that way lies exponential behaviour!

Reply:
> We don't do inlining based on usage analysis. 

Ah, yeah, I wrote that comment after reading the paper, but before reviewing the PR, so I did not realize we didn't have usage-based analysis.